File: src/studio/models/pipeline.py
"""
Pipeline DataFlow Model

Represents an agent orchestration pipeline in the Kaizen Studio platform.
"""

from studio.models import db


@db.model
class Pipeline:
    """
    Pipeline model for agent orchestration definitions.

    DataFlow generates these nodes automatically:
    - PipelineCreateNode
    - PipelineReadNode
    - PipelineUpdateNode
    - PipelineDeleteNode
    - PipelineListNode
    - PipelineCountNode
    - PipelineUpsertNode
    - PipelineBulkCreateNode
    - PipelineBulkUpdateNode
    - PipelineBulkDeleteNode
    - PipelineBulkUpsertNode
    """

    # Primary key - MUST be 'id' for DataFlow
    id: str

    # Organization and workspace reference
    organization_id: str
    workspace_id: str

    # Pipeline details
    name: str
    description: str  # Empty string if not provided

    # Orchestration pattern: sequential, parallel, router, supervisor, ensemble
    pattern: str

    # Status: draft, active, archived
    status: str

    # Audit fields
    created_by: str

    # Timestamps (ISO 8601 strings for DataFlow)
    created_at: str
    updated_at: str

File: apps/web/src/features/pipelines/types/pipeline.ts
/**
 * Pipeline type definitions for Kaizen Studio
 */

export interface Pipeline {
  id: string;
  name: string;
  description: string;
  pattern: PipelinePattern;
  nodes: PipelineNode[];
  edges: PipelineEdge[];
  status: "draft" | "active" | "archived";
  organization_id: string;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export type PipelinePattern =
  | "sequential"
  | "supervisor-worker"
  | "router"
  | "ensemble"
  | "parallel"
  | "hierarchical"
  | "pipeline"
  | "blackboard"
  | "reactive";

export interface PipelineNode {
  id: string;
  type: NodeType;
  position: { x: number; y: number };
  data: NodeData;
}

export type NodeType =
  | "agent"
  | "supervisor"
  | "router"
  | "synthesizer"
  | "connector"
  | "input"
  | "output";

export interface NodeData {
  label: string;
  agentId?: string;
  config?: Record<string, unknown>;
}

export interface PipelineEdge {
  id: string;
  source: string;
  target: string;
  sourceHandle?: string;
  targetHandle?: string;
  label?: string;
  condition?: string;
}

export interface PatternTemplate {
  id: string;
  name: string;
  pattern: PipelinePattern;
  description: string;
  nodes: PipelineNode[];
  edges: PipelineEdge[];
}
